public/app/core/components/Select/FolderPicker.tsx

```diff
@@ -0,0 +1,178 @@
+    const options: Array<SelectableValue<number>> = searchHits.map(hit => ({ label: hit.title, value: hit.id }));
+    if (contextSrv.isEditor && rootName?.toLowerCase().startsWith(query.toLowerCase())) {
+      options.unshift({ label: rootName, value: 0 });
+    }
+
+    if (enableReset && query === '' && initialTitle !== '') {
+      options.unshift({ label: initialTitle, value: undefined });
+    }
+
+    return options;
+  };
+
+  onFolderChange = async (newFolder: SelectableValue<number>) => {
+    if (!newFolder) {
+      newFolder = { value: 0, label: this.props.rootName };
+    }
+
+    this.setState(
+      {
+        folder: newFolder,
+      },
+      () => this.props.onChange({ id: newFolder.value!, title: newFolder.label! })
+    );
+  };
+
+  createNewFolder = async (folderName: string) => {
+    const newFolder = await getBackendSrv().createFolder({ title: folderName });
+    let folder = { value: -1, label: 'Not created' };
+    if (newFolder.id > -1) {
+      appEvents.emit(AppEvents.alertSuccess, ['Folder Created', 'OK']);
+      folder = { value: newFolder.id, label: newFolder.title };
+      await this.onFolderChange(folder);
+    } else {
+      appEvents.emit(AppEvents.alertError, ['Folder could not be created']);
+    }
+
+    return folder;
+  };
+
+  private loadInitialValue = async () => {
+    const { initialTitle, rootName, initialFolderId, enableReset, dashboardId } = this.props;
+    const resetFolder: SelectableValue<number> = { label: initialTitle, value: undefined };
+    const rootFolder: SelectableValue<number> = { label: rootName, value: 0 };
+
+    const options = await this.getOptions('');
+
+    let folder: SelectableValue<number> = { value: -1 };
+    if (initialFolderId || (initialFolderId && initialFolderId > -1)) {
+      folder = options.find(option => option.value === initialFolderId) || { value: -1 };
+    } else if (enableReset && initialTitle && initialFolderId === undefined) {
+      folder = resetFolder;
+    }
+
+    if (!folder) {
+      if (contextSrv.isEditor) {
+        folder = rootFolder;
+      } else {
+        // We shouldn't assign a random folder without the user actively choosing it on a persisted dashboard
+        const isPersistedDashBoard = !!dashboardId;
+        if (isPersistedDashBoard) {
+          folder = resetFolder;
+        } else {
+          folder = options.length > 0 ? options[0] : resetFolder;
+        }
+      }
+    }
+
+    this.setState(
+      {
+        folder,
+      },
+      () => {
+        // if this is not the same as our initial value notify parent
+        if (folder.value !== initialFolderId) {
+          this.props.onChange({ id: folder.value!, title: folder.text });
+        }
+      }
+    );
+  };
+
+  render() {
+    const { folder, validationError, hasValidationError } = this.state;
+    const { enableCreateNew } = this.props;
+
+    return (
+      <>
+        <div className="gf-form-inline">
```

Comment:
Next step here is to replace the legacy form styles with `<Forms.Field` component. Let's do it when migrating views with Folder Picker applied.